keps/sig-scheduling/1923-try-nominated-node-first/README.md

```diff
@@ -0,0 +1,224 @@
+for the incoming Pod to be scheduled.
+
+This KEP proposes to validate the availability of the nominated node in the
+new scheduling cycle before looping all over the nodes in the cluster for the
+phase of the filtering.
+
+## Motivation
+
+In real production environment, pods can have different priorites due to business
+needs, the preemption could happen to make sure high priority pods could got scheduled
+and spawned.
+
+For the cluster with huge number of computing nodes, looping over all the nodes in
+the cluster for the phase of the filtering is an unnegligible consuming for scheduler.
+
+### Goals
+
+- In the case of `pod.Status.NominatedNodeName` is set for the incoming pod, scheduler
+  will try the nominated node first for the filtering phase, failed to schedule on the
+  nominated node will fall back to verify all the nodes in the cluster instead of preemption
+  again.
+
+## Proposal
+
+### User Stories (Optional)
+
+As a premium user, I want my high priority application got spin up more quickly,
+I don't care where the application is running, whether there is any other nodes
+in the cluster have more disk space, memory size etc., as long as my application
+run normally.
+
+### Notes/Constraints/Caveats (Optional)
+
+When this feature is enabled the preemptor Pod might not be dispatched to the best
+candidated node in some corner case, e.g. another node releases the resources and
+becomes the best candidate while the victim pods got removed from the nominated node.
+
+## Design Details
+
+This feature hopes to improve the performance with the minimal change on the existing scheduling
+framework, scheduler will short-circuit to check the nominated node directly instead of looping
+all the nodes in the cluster, and if the nominated node doesn't fit, a fresh new scheduling cycle
+will be started to get an new node for scheduling.
+
+In most situation, the new scheduling cycle is not needed, but we still got the case like the
+topology changing, victim pods deletion is still on-the-fly in the following scheduling cycle.
+Those cases will make the nominated node not fits for the coming pod anymore, and a new scheduling
+cycle is needed in those case. This is no harm since it will eventually get convergent.
+
+### Implementation Details
+
+1. In filtering phase, which is currently implemented in the method of `findNodesThatFitPod`,
+   check the nominated directly if the incoming pod has the `pod.Status.NominatedNodeName` defined
+   and the feature gate is enabled.
+
+2. In case the nominated node doesn't suit for the incoming pod anymore, return `ErrNominateNode`
+   instead of `core.FitError`, because this will give scheduler a chance to clean up the nominated
+   node from the pod and find a new node to schedule instead of preemption again (might already have
+   another node available for scheduling during the period). A fresh new scheduling cycle will be
+   started later.
```

Comment:
are you suggesting that the rest of the nodes will not be evaluated in the same cycle? if so, why?

Reply:
I changed my mind to evaluate the rest of the nodes, this paragraph will be updated shortly.

Reply:
thanks, was this updated? I can't tell because the new patch was an amend.

Reply:
Keep in mind that removing the nominated node on the first failure is a source for other race conditions. For that reason, I suggest that all nodes should be evaluated and "new" nominated node calculated in the same cycle (if pod still doesn't fit).

Reply:
right, and I am not sure we need to clear the nominatednodename if we successfully bind the pod to a node. We need to clear it only if we fail to find a node that fits and preemption fails to suggest a new one.

Reply:
sorry for delay, this is not updated by far, I am still thinking about @alculquicondor 's suggestion here.  I am inclined to follow his suggestion now.

Reply:
I am going to keep the `nominatedNode` untouched, there is a potential issue around the  `nominatedNode` , please see: https://github.com/kubernetes/kubernetes/issues/95752, this is the issue that we have today, and should be a separated topic instead.  I tracked the issue there and assign the issue to myself, so that we won't mess those issues together.